sdks/python/apache_beam/dataframe/convert.py

```diff
@@ -16,13 +16,23 @@
 
 from __future__ import absolute_import
 
+import typing
+
 import inspect
 
 from apache_beam import pvalue
 from apache_beam.dataframe import expressions
 from apache_beam.dataframe import frame_base
 from apache_beam.dataframe import transforms
 
+if typing.TYPE_CHECKING:
+  # pylint: disable=ungrouped-imports
+  from typing import Any
+  from typing import Dict
+  from typing import Tuple
+  from typing import Union
```

Comment:
Scoping these imports is not a bad idea since it keeps the module namespace cleaner, but there are a couple of issues with this:

- once we get to python 3.x, and move from type comments to annotations, this will fail, as we'll be referencing non-existent objects in our annotations.  
- it's not consistent with how we've done this throughout the rest of the code

Reply:
I should mention that a solution to the first issue is that we can refer to these types as strings, such as `'Any'`, but that's certainly a lot more awkward, and developers are likely to forget to do so and get confused/frustrated.

Reply:
I actually prefer unconditionally importing them, but was just trying to avoid lint issues (and did see this pattern elsewhere). Changed.

Reply:
So lint complains about unguarded imports, so I put them back. We'll just to a massive sweep to fix these when we change to use type annotations.

Reply:
What's the lint error?   Is it because of the unused `typing` import?  

I'm confused because unguarded typing imports are used all over the beam codebase without any lint errors.  Check `pipeline`, `pipeline_context`, `pipeline_options`, for starters.

Reply:
I had a look at the lint errors, and they are legitimate, but scoping the imports is not the right solution.

For example, these:

```
  | apache_beam/dataframe/frame_base.py:21:0: W0611: Unused Any imported from typing (unused-import)
  | apache_beam/dataframe/frame_base.py:23:0: W0611: Unused Tuple imported from typing (unused-import) 
  | apache_beam/dataframe/frame_base.py:24:0: W0611: Unused Union imported from typing (unused-import)
```

`Any`, `Tuple`, and `Union` _aren't_ used in this module. `Dict` is the only typing import that is used.

Reply:
Note that pylint is pretty good about detecting the use of typing imports, even in comments.  I have encountered a few edge cases but I reported them and they were fixed pretty quickly.   

The only issue I know of in the version of pylint that we're using is with `typing.BinaryIO`.  You'll notice a few lines like these in the code:

```python
from typing import BinaryIO  # pylint: disable=unused-import
```

I [reported](https://github.com/PyCQA/pylint/issues/3217) this bug and it's been fixed, so we can actually get rid of these exceptions if we upgrade to the latest pylint.

Reply:
This note hasn't been addressed:  

> I had a look at the lint errors, and they are legitimate, but scoping the imports is not the right solution.

see above.

Reply:
Oh, yes. Thanks for catching this. Fixed.

Reply:
I think it still needs to be fixed for `dataframe.convert`.

Reply:
[sigh] It still didn't like PCollection. `apache_beam/dataframe/transforms.py:30:0: W0611: Unused PCollection imported from apache_beam.pvalue (unused-import)`. But the rest are OK.

Reply:
And pandas needs to be guarded. Hopefully that should be it. PTAL.